src/jemalloc/include/jemalloc/internal/jemalloc_internal.h.in

```diff
@@ -1041,7 +1041,7 @@ iralloct(void *ptr, size_t size, size_t extra, size_t alignment, bool zero,
 			}
 		}
 
-		return NULL;
+		return (ret != NULL) ? (ret) : (ptr);
```

Comment:
The whole story was started by some failing test and the test engineer surprised with the fact that you can i.e. reallocate from 256B to 16B and it always (?) succeeds, but at the same time, you can't reallocate from 4M to 16B.
Since we know the jemalloc internals, it's clear why it happens, but for the end user it might be surprising - "I have 4M block, but need only 16B of it - why could realloc fail"? (assuming single-threaded program)
Of course, we can say it's because of the internal design of the allocator, and when you're close to the 100% heap utilization (which means - no more free chunks), then there is no guarantee that malloc/realloc succeeds, even if (theoretically) you still have enough free space and/or you attempt to shrink the allocated object.

On the other hand, if realloc may do nothing and return the pointer to the old (larger) object, it doesn't differ much from returning NULL. The old object/pointer is still valid and its size is unchanged. The difference is that with NULL, the user got the clear notification on that fact, so they know the object has not been shrinked actually. With my patch, they might be confused.

After all, I think I'll discard this patch.